src/pcap/client.go

```diff
@@ -0,0 +1,335 @@
+package pcap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"code.cloudfoundry.org/bytefmt"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/google/gopacket/pcapgo"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+)
+
+const logProgressWait = 5 * time.Second
+
+type MessageWriter interface {
+	WriteMessage(message *Message)
+}
+
+type LogMessageWriter struct {
+	Log *zap.Logger
+}
+
+// WriteMessage accepts a Message and writes a log-line using a log-level corresponding to the severity of the message.
+func (c LogMessageWriter) WriteMessage(message *Message) {
+	formattedMessage := fmt.Sprintf("%s(%s): %s", message.Origin, message.Type, message.Message)
+	c.Log.Log(MessageLogLevel(message), formattedMessage)
+}
+
+// MessageLogLevel translates the message types to appropropriate default log levels.
+func MessageLogLevel(message *Message) zapcore.Level {
+	switch message.Type {
+	case MessageType_UNKNOWN:
+		return zapcore.WarnLevel
+	case MessageType_INSTANCE_UNAVAILABLE:
+		return zapcore.WarnLevel
+	case MessageType_START_CAPTURE_FAILED:
+		return zapcore.WarnLevel
+	case MessageType_INVALID_REQUEST:
+		return zapcore.ErrorLevel
+	case MessageType_CONGESTED:
+		return zapcore.WarnLevel
+	case MessageType_LIMIT_REACHED:
+		return zapcore.WarnLevel
+	case MessageType_CAPTURE_STOPPED:
+		return zapcore.InfoLevel
+	case MessageType_CONNECTION_ERROR:
+		return zapcore.ErrorLevel
+	}
+	return zapcore.ErrorLevel
+}
+
+// Client provides a reusable client for issuing capture requests against the pcap-api.
+type Client struct {
+	packetFile    *os.File
+	log           *zap.Logger
+	stream        API_CaptureClient
+	messageWriter MessageWriter
+	stopped       bool
+	aPIClient
+}
+
+func CloseQuietly(closer io.Closer) {
+	_ = closer.Close()
+}
+
+// NewClient sets up logging for the client and creates the outputFile.
+// It assumes that the outputFile does not pre-exist and that the path is writeable (should be checked by CLI).
+//
+// NewClient returns a new Client if there are no issues with outputFile creation.
+func NewClient(outputFile string, logger *zap.Logger, writer MessageWriter) (*Client, error) {
+	var err error
+
+	client := &Client{log: logger, messageWriter: writer}
+
+	if len(outputFile) == 0 {
+		client.packetFile = os.Stdout
+	} else {
+		client.packetFile, err = os.Create(outputFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+func (c *Client) Stop() {
+	c.StopRequest()
+}
+
+func (c *Client) Wait() {
+	// waiting is implicit. Needed for fulfulling the pcap.Stoppable interface.
+}
```

Comment:
This seems very wrong, it makes a promise that it cannot fulfill. If the interface is too strict we should loosen it or make the client support this functionality.

Reply:
I will instead remove the Stoppable interface from the client, as it doesn't fit there.